test/e2e/etcd_multi_node_test.go

```diff
@@ -0,0 +1,414 @@
+			),
+		),
+	).To(Succeed())
+}
+
+// checkJobReady checks k8s job associated pod is ready, up and running.
+// checkJobReady is a helper function, uses Gomega Eventually.
+func checkJobReady(ctx context.Context, cl client.Client, jobName string) {
+	r, _ := k8s_labels.NewRequirement("job-name", selection.Equals, []string{jobName})
+	opts := &client.ListOptions{
+		LabelSelector: k8s_labels.NewSelector().Add(*r),
+		Namespace:     namespace,
+	}
+
+	EventuallyWithOffset(1, func() error {
+		podList := &corev1.PodList{}
+		if err := cl.List(ctx, podList, opts); err != nil {
+			return fmt.Errorf("error occurred while getting pod object: %v ", err)
+		}
+
+		if len(podList.Items) == 0 {
+			return fmt.Errorf("job %s associated pod is not scheduled", jobName)
+		}
+
+		for _, c := range podList.Items[0].Status.Conditions {
+			if c.Type == corev1.PodReady && c.Status == corev1.ConditionTrue {
+				return nil
+			}
+
+		}
+		return fmt.Errorf("waiting for pod %v to be ready", podList.Items[0].Name)
+	}, timeout, pollingInterval).Should(BeNil())
+}
+
+// etcdZeroDownTimeValidatorJob returns k8s job which ensures
+// Etcd cluster zero down time by continuously checking etcd cluster health.
+// This job fails once health check fails and associated pod results in error status.
+func startEtcdZeroDownTimeValidatorJob(ctx context.Context, cl client.Client,
+	etcd *v1alpha1.Etcd, testName string) *batchv1.Job {
+	job := etcdZeroDownTimeValidatorJob(etcd.Name+"-client", testName, etcd.Spec.Etcd.ClientUrlTLS)
+
+	logger.Info(fmt.Sprintf("Creating job %s to ensure etcd zero downtime", job.Name))
+	ExpectWithOffset(1, cl.Create(ctx, job)).ShouldNot(HaveOccurred())
+
+	// Wait until zeroDownTimeValidator job is up and running.
+	checkJobReady(ctx, cl, job.Name)
+	logger.Info(fmt.Sprintf("Job %s is ready", job.Name))
+	return job
+}
+
+//getEtcdLeaderPodName returns the leader pod name by using lease
+func getEtcdLeaderPodName(ctx context.Context, cl client.Client, namespace string) (*types.NamespacedName, error) {
+	leaseList := &v1.LeaseList{}
+	opts := &client.ListOptions{Namespace: namespace}
+	ExpectWithOffset(1, cl.List(ctx, leaseList, opts)).ShouldNot(HaveOccurred())
+
+	for _, lease := range leaseList.Items {
+		if lease.Spec.HolderIdentity == nil {
+			return nil, fmt.Errorf("error occurred while finding leader, etcd lease %q spec.holderIdentity is nil", lease.Name)
+		}
+		if strings.Contains(*lease.Spec.HolderIdentity, "Leader") {
+			return &types.NamespacedName{Namespace: namespace, Name: lease.Name}, nil
+		}
+	}
+	return nil, fmt.Errorf("leader doesn't exist for namespace %q", namespace)
+}
+
+// getPodLogs returns logs for the given pod
+func getPodLogs(ctx context.Context, PodKey *types.NamespacedName, opts *corev1.PodLogOptions) (string, error) {
+	typedClient, err := getKubernetesTypedClient(kubeconfigPath)
+	if err != nil {
+		return "", err
+	}
+
+	req := typedClient.CoreV1().Pods(namespace).GetLogs(PodKey.Name, opts)
```

Comment:
depending upon pod logs is very error prone, instead it should be reflected in the status as a condition. So for now checking the logs is fine but this should be corrected.

Reply:
I agree with @unmarshall , so what I would suggest is get the db size of each member (before and after defragmentation call), now compare the db size, if new db size of each member is less then old db size then defragmentation was successful. check [here](https://github.com/gardener/etcd-backup-restore/blob/master/pkg/defragmentor/defrag_test.go#L297-L340)
My only fear is it might cause flaky tests as we are doing the similar thing in unit tests of defragmentation in backup-restore and it becomes flaky.

Reply:
@ishan16696 
How do we i know defragmention is completed or not ?
Is it based on db size ?

Reply:
yes, basically defragmentation freeup the database space but before calling the defragmentation you have to put some dummy data into etcd cluster, so that database get increase somewhat  and defragmentation will be effective.

Reply:
@ishan16696 
 I hope @unmarshall point is that we should update the status of defgramentation as one of the status conditions.
 Right now we rely on pod logs for degfragmention status and  I dont think so  checking the database size is a proper way of confirming the degferamention is completed or not.

Reply:
@unmarshall may be we can go ahead with what @seshachalam-yv is doing and may be I will try to pick and improve this. WDYT?

Reply:
ok as i indicated in the first comment, checking logs is fine for now, but this needs to be changed. let us discuss this separately.